src/builder.rs

```diff
@@ -111,6 +111,73 @@ impl<W: Write> Builder<W> {
         append(self.get_mut(), header, &mut data)
     }
 
+    /// Appends the contents of an existing archive to this archive.
+    ///
+    /// This function will iterate through the entries of the supplied archive,
+    /// appending them one by one to the target.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error for any intermittent I/O error which
+    /// occurs when either reading or writing.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::fs::File;
+    /// use tar::{Archive, Builder};
+    ///
+    /// let mut input = Archive::new(File::open("/path/to/existing.tar").unwrap());
+    ///
+    /// let mut output_file = File::create("/path/to/target.tar").unwrap();
+    /// let mut output = Builder::new(output_file);
+    ///
+    /// output.append_archive(&mut input).unwrap();
+    /// output.finish().unwrap();
+    /// ```
+    pub fn append_archive<R: Read>(&mut self, archive: &mut Archive<R>) -> io::Result<()> {
+        let entries = archive.entries()?;
+
+        for entry in entries {
+            let file_entry = entry?;
+            let header = file_entry.header().clone();
+            self.append(&header, file_entry)?;
+        }
+
+        Ok(())
+    }
+
+    /// Appends an entry from an existing archive to this archive.
+    ///
+    /// This function will append an existing `Entry` to the builder target.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error for any intermittent I/O error which
+    /// occurs when either reading or writing.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::fs::File;
+    /// use tar::{Archive, Builder, Entry};
+    ///
+    /// let mut input = Archive::new(File::open("/path/to/existing.tar").unwrap());
+    /// let mut entry = input.entries().unwrap().next().unwrap().unwrap();
+    ///
+    /// let mut output_file = File::create("/path/to/target.tar").unwrap();
+    /// let mut output = Builder::new(output_file);
+    ///
+    /// output.append_entry(&mut entry).unwrap();
+    /// output.finish().unwrap();
+    /// ```
+    pub fn append_entry<R: Read>(&mut self, entry: &mut Entry<R>) -> io::Result<()> {
+        let header = entry.header().clone();
```

Comment:
I'm a little worried that this loses information because it doesn't handle things like long path names and such?

Reply:
OK - can you explain this one a bit more? I see that in https://github.com/alexcrichton/tar-rs/blob/master/src/header.rs#L888 clone uses `self.bytes` - should I be checking `is_ustar` and `is_gnu` instead and constructing the appropriate `Header` instance?

Also could you describe an example that I could use as a test to verify this works as expected?

Cheers! G

Reply:
Oh sorry yeah, I should clarify! The `Entry<T>` type has a number of inherent methods:

* `Entry::path`
* `Entry::link_name`
* `Entry::path_extensions`

With various extensions to the `tar` format one instance of `Entry<T>` may actually be multiple tar-level entries in the underlying file. For example some entries just say "this is a longer path name than you could otherwise store". That's what the `raw(true)` for the entries iterator above is doing, it's avoiding automatically coalescing entries together and simply appending them one-by-one.

Here though an `Entry` may represent multiple tar-level entries that need to be emitted, but I believe the implementation here is only emitting one. Perhaps at least `path` could be handled? I think if you call `append_path` it should largely do the right thing.

Reply:
Right I think I understand. I see looking at the [UStar format](https://en.wikipedia.org/wiki/Tar_(computing)#UStar_format) that the `'g'` and `'x'` header type flags can provide all sorts of stuff:
> extended header with meta data for the next file in the archive

I had no idea tar files could be that exciting!  

I'll add in some test cases that will hopefully surface the shortcoming in the implementation and then fix it up.

Reply:
Yeah tar files ended up being full of surprises, I thought writing a parser library like this wouldn't be too bad and here we are!

Reply:
Yeah no kidding! 

I managed to replicate the issue with path truncation you mentioned, and have a simple change that handles long path names (as per your suggestion):
```self.append(&mut header, entry)?;``` -> ```self.append_data(&mut header, path, entry)?;```

The only issue I have is with a test assertion on the contents of the generated entry header - specifically with the checksum. It seems the `Builder` generates a different format checksum from `GNU tar`:

Builder | GNU tar
------------ | -------------
`0007666 `|`007666  `

i.e. the `Builder` creates a 7 digit representation followed by a space, while `GNU tar` creates a 6 digit representation followed by a NUL and then a space.

Wikipedia says:
> It is stored as a six digit octal number with leading zeroes followed by a NUL and then a space. Various implementations do not adhere to this format. For better compatibility, ignore leading and trailing whitespace, and take the first six digits. 

Do you think I should update the checksum implementation to generate `6 digits w/ leading zeros + NUL + space` so that it matches the `GNU tar` behaviour?

Reply:
More specifically, I'm referring to the use of `octal_into` for checksum values https://github.com/alexcrichton/tar-rs/blob/master/src/header.rs#L1392 - I think the last two array values should be `NUL` and `space` (`[0, 32]`).